picky/src/x509/pkcs7/authenticode.rs

```diff
@@ -0,0 +1,1886 @@
+
+        // 7. Given the x509 certificate, compare SignerInfo::encrypted_digest against hash of authenticated attributes and hash of ContentInfo
+        let public_key = signing_certificate.public_key();
+
+        let hash_algo = ShaVariant::try_from(Into::<ObjectIdentifierAsn1>::into(digest_algorithm.oid().clone()))
+            .map_err(AuthenticodeError::UnsupportedAlgorithmError)?;
+
+        let authenticated_attributes = self.authenticode_signature.authenticated_attributes();
+        let mut raw_attributes = picky_asn1_der::to_vec(authenticated_attributes)?;
+        // According to the RFC:
+        //
+        // "[...] The Attributes value's tag is SET OF, and the DER encoding ofs
+        // the SET OF tag, rather than of the IMPLICIT [0] tag [...]"
+        raw_attributes[0] = Tag::SET.inner();
+
+        let signature_algorithm_identifier = AlgorithmIdentifier::new_rsa_encryption_with_sha(hash_algo)
+            .map_err(AuthenticodeError::UnsupportedAlgorithmError)?;
+        let signature_algorithm = SignatureAlgorithm::from_algorithm_identifier(&signature_algorithm_identifier)?;
+        signature_algorithm
+            .verify(public_key, &raw_attributes, &signer_info.signature.0 .0)
+            .map_err(AuthenticodeError::SignatureError)?;
+
+        // 8. PKCS9_MESSAGE_DIGEST attribute exists and that its value matches hash of ContentInfo.
+        let message_digest_attr = authenticated_attributes
+            .iter()
+            .find(|attr| matches!(attr.value, AttributeValues::MessageDigest(_)))
+            .ok_or(AuthenticodeError::NoMessageDigest)?;
+
+        let hash_algo = HashAlgorithm::try_from(hash_algo).map_err(AuthenticodeError::UnsupportedHashAlgorithmError)?;
+
+        let mut raw_spc_indirect_data_content = picky_asn1_der::to_vec(&spc_indirect_data_content.data)?;
+        let mut raw_message_digest = picky_asn1_der::to_vec(&spc_indirect_data_content.message_digest)?;
+        raw_spc_indirect_data_content.append(&mut raw_message_digest);
+
+        let content_info_hash = hash_algo.digest(&raw_spc_indirect_data_content);
+
+        if let AttributeValues::MessageDigest(message_digest_attr_val) = &message_digest_attr.value {
+            if message_digest_attr_val
+                .0
+                .first()
+                .expect("At least one element is always present in Asn1SetOf AttributeValues")
+                != &content_info_hash
+            {
+                return Err(AuthenticodeError::HashMismatch);
+            }
+        }
+
+        // 9. The signing certificate must contain either the extended key usage (EKU) value for code signing,
+        // or the entire certificate chain must contain no EKUs
+        let certificates = self.authenticode_signature.0.certificates();
+        check_eku_code_signing(&certificates)?;
+
+        Ok(())
+    }
+
+    fn verify_signing_certificate(&self) -> AuthenticodeResult<()> {
```

Comment:
Picky name convention for helpers
```suggestion
    fn h_verify_signing_certificate(&self) -> AuthenticodeResult<()> {
```